IMT 2200 - Introducción a Ciencia de Datos<br>
**Pontificia Universidad Católica de Chile**<br>
**Instituto de Ingeniería Matemática y Computacional**<br>
**Semestre 2025-S2**<br>
**Profesor:** Rodrigo A. Carrasco <br>

---

## Tarea 01 – Cargando y Analizando Datos

- **Fecha de Entrega:** martes 26 de agosto de 2025, a las 23:59.
- 
**Formato de entrega:** Notebook ejecutado y comentado (`.ipynb`) en l emódulo de Tara 01 habilitado en Canvas.




## Instrucciones

- Esta Tarea debe desarrollarse de manera totalmente *individual*, de acuerdo a lo establecido en la sección de Integridad Académica en el programa del curso.
- La Tarea debe ser desarrollada en lenguaje de programación Python y la entrega en formato Jupyter Notebook.
- El desarrollo del Notebook debe ser claro y ordenado, incluyendo anotaciones (markdown) y comentarios que permitan seguir fácilmente el código y los pasos implementados a los correctores, y siguiendo buenas prácticas de programación. La presentación y claridad del notebook y código forman parte de la evaluación de la tarea.
- Notebook **autocontenible** que:
   - Ejecute sin errores todas las celdas.
  - Contenga tanto el código como los comentarios y explicaciones necesarias.
  - Incluya visualizaciones claras y correctamente etiquetadas.
- No se aceptarán notebooks con celdas rotas o que dependan de rutas externas no indicadas en la tara.

- Deben hacer sus consultas y comentarios sobre la Tarea a traves del canal de Tareas en eo del curso en Canvas.os.
cteriza.

## 1. Objetivos

- Aplicar los conceptos iniciales de manejo de datos y análisis exploratorio vistos en clases.
- Practicar la lectura, limpieza y manipulación de datos en Python.
- Desarrollar habilidades para visualizar y describir patrones y tendencias en conjuntos de datos reales.
- Fomentar la capacidad de comunicar resultados de forma clara y fundamentada.

### 1.1 Objetivo educacional

Esta Tarea tiene como objetivo que los estudiantes desarrollen la capacidad de manejar algunas de las librerías centrales para el desarrollo de Ciencia de Datos, con foco en la lectura y exploración de datos. 

Para los ejercicios a continuación, usted deberá leer, inspeccionar, manipular y graficar conjuntos de datos en distintos formatos, de manera de responder las preguntas de cada parte de la Tarea.

### 1.2 Pregunta de ciencia de datos

Para esta tarea, tendremos como objetivo comprender cómo han cambiado los juegos de mesa en los últimos 40 años. Específicamente queremos saber qué tipos de juegos se han vuelto más comunes hoy en día y qué los caracteriza.

### 1.3 Recomendaciones
- Utiliza las librerías sugeridas en el notebook o justifica brevemente si incorporas otras.
- Revisa que todas las celdas se ejecuten en orden, desde el inicio, sin errores.
- Comenta tu código para explicar qué hace cada sección relevante.
- Asegúrate de que las visualizaciones sean fáciles de interpretar y tengan títulos y etiquetas adecuados.

## 2. Datos

Estaremos utilizando información extraída desde [BoardGameGeek.com](https://boardgamegeek.com/), una plataforma para aficionados de los juegos de mesa, que permite a sus usuarios registrar, calificar e intercambiar sus juegos favoritos. Actualmente, BGG es una de las bases de datos más extensa y diversa de juegos de mesa.

El dataset con el que trabajaremos consiste en un grupo de archivos CSV que contienen información sobre más de 100.000 juegos de mesa almacenados en la plataforma. Este puede descargarse directamente desde el siguiente enlace: https://www.kaggle.com/datasets/mshepherd/board-games Para descargar los datos, haga click en el botón de **Download**, donde podrá descargar los archivos como `.zip`, o bien utilizar la API de Kaggle.

Para el desarrollo de esta tarea, solo utilizaremos los archivos con el prefijo `bgg_`.

**Si utiliza la API de Kaggle para descargar los datos, deje el código utilizado en la siguiente celda:**

In [ ]:
"""
# Por falta de tiempo no podido terminar la descarga mediante la API de Kaggle, logre descargar los datos como csv, mas algunos de estos
# se muestran comprimidos en zip, por lo que no se pueden leer directamente con pandas.

# Procedere con la descarga manual de los archivos csv y zip, y los descomprimire en la carpeta Data.

# Descarga de datos desde Kaggle usando el codigo facilitado para descargar datasets
# %pip install kagglehub[pandas-datasets]

import kagglehub
from kagglehub import KaggleDatasetAdapter
from pathlib import Path

# Carpeta de salida
out_dir = Path("Data")
out_dir.mkdir(exist_ok=True)

# Archivos disponibles en el dataset (iniciados con bgg_)
files = [
    "bgg_GameItem.csv",
    "bgg_Category.csv",
    "bgg_GameFamily.csv",
    "bgg_GameType.csv",
    "bgg_Mechanic.csv",
    "bgg_Person.csv",
    "bgg_Publisher.csv",
]
# Descargar cada archivo
for file_path in files:
    print(f"\nDescargando {file_path} ...")
    
    # KaggleHub devuelve la ruta local del archivo en la caché
    local_path = kagglehub.dataset_download("mshepherd/board-games", file_path)
    
    # Leer bytes del archivo original
    with open(local_path, "rb") as f:
        data = f.read()

    # Guardar en Data/
    out_file = out_dir / file_path
    with open(out_file, "wb") as f:
        f.write(data)
    
    print(f"Guardado en {out_file.resolve()} ({out_file.stat().st_size/1024:.2f} KB)")

print("\nDescarga completada.")

"""


Descargando bgg_GameItem.csv ...
Guardado en C:\Users\carlo\Desktop\PUC\IMT2200 - Intro a Ciencia de Datos\Desarrollo\tarea 01\Data\bgg_GameItem.csv (15700.02 KB)

Descargando bgg_Category.csv ...
Guardado en C:\Users\carlo\Desktop\PUC\IMT2200 - Intro a Ciencia de Datos\Desarrollo\tarea 01\Data\bgg_Category.csv (1.50 KB)

Descargando bgg_GameFamily.csv ...
Guardado en C:\Users\carlo\Desktop\PUC\IMT2200 - Intro a Ciencia de Datos\Desarrollo\tarea 01\Data\bgg_GameFamily.csv (165.07 KB)

Descargando bgg_GameType.csv ...
Guardado en C:\Users\carlo\Desktop\PUC\IMT2200 - Intro a Ciencia de Datos\Desarrollo\tarea 01\Data\bgg_GameType.csv (0.21 KB)

Descargando bgg_Mechanic.csv ...
Guardado en C:\Users\carlo\Desktop\PUC\IMT2200 - Intro a Ciencia de Datos\Desarrollo\tarea 01\Data\bgg_Mechanic.csv (4.32 KB)

Descargando bgg_Person.csv ...
Guardado en C:\Users\carlo\Desktop\PUC\IMT2200 - Intro a Ciencia de Datos\Desarrollo\tarea 01\Data\bgg_Person.csv (1134.62 KB)

Descargando bgg_Publisher.csv 

### 2.1 Librerías

Para esta tarea recomendamos al menos usar las librerías indicadas en la siguiente celda del Notebook. Puede agregar otras si lo estima conveniente para responder de mejor forma las preguntas de la Tarea.

In [14]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

## 3. Desarrollo

Para cada una de las siguientes preguntas o actividades incluya una o más celdas de código y Markdown con las respuestas o comentarios necesarios para contestar la pregunta o lograr la actividad. Agregue sus celdas a continuación de cada pregunta para que el Notebook quede ordenado.

En el caso de gráficos, figuras o tablas, asegúrese que todas tengan título, etiquetas en los ejes o haya claridad de los contenidos.

### 3.1 DataFrame unificado (1 punto)

Vamos a cargar en un DataFrame los datos de juegos guardados en el documento `bgg_GameItem.csv`. Al inspeccionar nuestros datos, podemos notar que los valores de ciertas comunas vienen "codificados" con ID. Para comenzar, vamos a juntar la información de los distintos archivos descargados.

**a) (0.8 pts)** Genere un DataFrame único con toda la información de cada juego, incluyendo: nombres de artistas y diseñadores involucrados, mecánicas, categorías, tipo, familia y editorial. Guarde este DataFrame en un nuevo archivo CSV.

In [ ]:
#cargamos los archivos con pandas

game_item = pd.read_csv("Data/bgg_GameItem.csv")
category = pd.read_csv("Data/bgg_Category.csv")
game_family = pd.read_csv("Data/bgg_GameFamily.csv")
game_type = pd.read_csv("Data/bgg_GameType.csv")
mechanic = pd.read_csv("Data/bgg_Mechanic.csv")
person = pd.read_csv("Data/bgg_Person.csv")
publisher = pd.read_csv("Data/bgg_Publisher.csv")

#creo esta funcion para hacer el merge de los dataframes, para juntar los nombres a los ids
def decode_column(base_df, col_ids, ref_df):
    temp = base_df[["name", col_ids]].copy()
    temp[col_ids] = temp[col_ids].str.split(",")
    temp = temp.explode(col_ids)
    # convertir a numérico para evitar problemas de tipo
    temp[col_ids] = pd.to_numeric(temp[col_ids])
    temp = temp.merge(ref_df, left_on=col_ids, right_on="bgg_id", how="left")
    # Agrupar por el nombre del juego y agregar los nombres relacionados en una lista
    return temp.groupby("bgg_id")["bgg_id"].apply(list).reset_index()

#aplicamos la funcion a las columnas que contienen ids
artists_df   = decode_column(game_item, "artist", person)
designers_df = decode_column(game_item, "designer", person)
mechanics_df = decode_column(game_item, "mechanic", mechanic)
categories_df = decode_column(game_item, "category", category)
types_df = decode_column(game_item, "game_type", game_type)
families_df = decode_column(game_item, "family", game_family)
publishers_df = decode_column(game_item, "publisher", publisher)

#juntamos todos los dataframes en uno solo
df = game_item[["name", "year"]].copy()
df = df.merge(artists_df, on="name", how="left")
df = df.merge(designers_df, on="name", how="left")
df = df.merge(mechanics_df, on="name", how="left")
df = df.merge(categories_df, on="name", how="left")
df = df.merge(types_df, on="name", how="left")
df = df.merge(families_df, on="name", how="left")
df = df.merge(publishers_df, on="name", how="left")

df.head(10)

KeyError: 'Column not found: name'

**b) (0.2 pts)** ¿Cuánto espacio en disco ocupa este DataFrame? ¿Cuánto espacio en disco ocupan los documentos CSV separados? ¿A qué se debe esta diferencia? Comente.



### 3.2 Juegos publicados anualmente (1.5 puntos)


**a) (0.5 pts)** Limpie los datos para dejar solo aquellos que tienen valores válidos de año de publicación. Luego responda: ¿cuál es el rango de años con el que estamos trabajando? ¿Tienen sentido estos años?


**b) (0.5 pts)** Seleccione solamente los juegos entre los años 1980 y 2025. Luego grafique la cantidad de juegos publicados por año. ¿Cómo es esta tendencia? Comente.

**c) (0.5 pts)** ¿Entre qué años hubo un mayor aumento de publicación de juegos de mesa según los registros de BGG?

### 3.3 Análisis de duración y complejidad (1.5 puntos)

Si bien hay muchas posibles características que podemos explorar para los juegos de mesa, esta vez nos centraremos en el tiempo de juego y la complejidad. En esta sección queremos comprender si se ha modificado notablemente la duración promedio y la percepción de complejidad de los juegos de mesa a través de los años.

**a) (0.3 pts)** Inspecciones y filtre los datos que tengan valores válidos para: tiempo mínimo de juego, tiempo máximo de juego y complejidad. En el caso de que existan "outliers", puede descartarlos, cosiderando un rango razonable para las variables anteriores. Justifique su desición.

**b) (0.2 pts)** Veremos ahora si ha cambiado la duración promedio de los juegos de mesa en nuestro rango de años seleccionado. Primero, cree una nueva columna `avg_time` en el dataset con la duración promedio supuesta para cada juego.

**c) (0.5 pts)** Grafique la duración promedio de juegos al año entre 1980 y 2025. ¿Existe alguna tendencia? Asegúrese de colocar las unidades correspondientes en sus ejes si es necesario. Comente sus resultados.

**d) (0.5 pts)** Nos interesa saber si los juegos de mesa recientes son más complejos que aquellos publicados antes de los 2000. BoardGameGeek permite evaluar la complejidad (o "weight") de un juego en una escala de 1 a 5, siendo 1 un juego "liviano" o fácil de entender, y 5 un juego "pesado" o complejo. Primero, seleccione los juegos que han sido evaluados por al menos 100 usuarios. Luego grafique la complejidad promedio de los juegos según año. Responda: ¿ha cambiado la percepción de complejidad entre juegos entre 1980 y la actualidad?

### 3.4 Análisis de categorías comunes (2 puntos)

Existe una gran diversidad de categorías de juegos de mesa. Ahora nos concentraremos en un grupo específico de ellas, con el fin de analizar cómo a cambiado la cantidad de juegos de estas clases desde 1980 hasta hoy.

**a) (0.7 pts)** ¿Cuáles son las 5 categorías más comunes en los juegos del dataset? Muestre la cantidad de juegos que hay de cada una. Puede graficar estos valores, o bien, entregar un DataFrame con sus valores.

**b) (0.3 pts)** Para cada una de estas 5 categorías, cree un DataFrame que contenga la cantidad de juegos en el dataset según año. Luego junte estos DataFrames en uno solo con los atributos de "Año", "Categoría" y "Cantidad".

**c) (1 pto)** Grafique, en un solo gráfico y con distintos colores, la cantidad de juegos por año según categoría. Preocúpese de ponerle etiquetas al gráfico para identificar cada categoría y una leyenda donde se muestre cada una. Comente: ¿ha habido un cambio entre los juegos más comunes en los años 80 y hoy?


### 3.5 Análisis Crítico (Bono +0.5 puntos)

¿Qué limitaciones o problemas encontraste en los datos?